src/Microsoft.AspNetCore.SignalR.Client.Core/HubConnection.cs

```diff
@@ -788,19 +781,67 @@ private Task WaitConnectionLockAsync([CallerMemberName] string memberName = null
         private class Subscription : IDisposable
         {
             private readonly InvocationHandler _handler;
-            private readonly List<InvocationHandler> _handlerList;
+            private readonly InvocationHandlerList _handlerList;
 
-            public Subscription(InvocationHandler handler, List<InvocationHandler> handlerList)
+            public Subscription(InvocationHandler handler, InvocationHandlerList handlerList)
             {
                 _handler = handler;
                 _handlerList = handlerList;
             }
 
             public void Dispose()
             {
-                lock (_handlerList)
+                _handlerList.Remove(_handler);
+            }
+        }
+
+        private class InvocationHandlerList
+        {
+            private readonly List<InvocationHandler> _invocationHandlers;
+            // A lazy cached copy of the handlers that doesn't change for thread safety. 
+            // Adding or removing a handler sets this to null.
+            private InvocationHandler[] _copiedHandlers;
+
+            internal InvocationHandlerList(InvocationHandler handler)
+            {
+                _invocationHandlers = new List<InvocationHandler>() { handler };
+            }
+
+            internal InvocationHandler[] GetCopiedHandlers()
```

Comment:
nit: Just name it `GetHandlers`. I don't like the name `CopiedHandlers` and want to keep it as private as possible ;P